File: sdks/python/src/etrace/langchain.py
"""Callback adapter for framework-driven agent tracing.

Translates callback events into the generic :class:`~etrace.tracing.RunTracker`
API so that LLM calls, tool executions, and chain steps become properly nested
etrace spans.

Usage::

    from etrace.langchain import EtraceLangChainHandler

    handler = EtraceLangChainHandler()
    result = agent.invoke(
        {"messages": [HumanMessage(content="...")]},
        config={"callbacks": [handler]},
    )

The handler is auto-registered when the supported callback package is installed
and ``etrace.init()`` is called with callback integration enabled.
"""

from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING, Any

from .tracing import RunTracker, _Run

logger = logging.getLogger("etrace.langchain")

if TYPE_CHECKING:
    from uuid import UUID

try:
    from langchain_core.callbacks import BaseCallbackHandler  # type: ignore[import-not-found]

    _HAS_LANGCHAIN = True
except ImportError:
    # Provide a stub so the class can still be imported for type checking
    class BaseCallbackHandler:  # type: ignore[no-redef]
        pass

    _HAS_LANGCHAIN = False


def _serialize_messages(messages: list[Any]) -> list[dict[str, Any]]:
    """Convert callback message objects to a serializable list."""
    result = []
    for msg in messages:
        entry = {
            "role": getattr(msg, "type", "unknown"),
            "content": getattr(msg, "content", ""),
        }
        tc = getattr(msg, "tool_calls", None)
        if tc:
            entry["tool_calls"] = [{"name": t.get("name"), "args": t.get("args")} for t in tc]
        result.append(entry)
    return result


class EtraceLangChainHandler(BaseCallbackHandler):  # type: ignore[misc]
    """Callback handler backed by :class:`RunTracker`.

    Maps callback ``run_id`` / ``parent_run_id`` values to tracker run IDs so
    spans nest correctly.
    """

    def __init__(self) -> None:
        if not _HAS_LANGCHAIN:
            raise ImportError(
                "langchain-core is required for EtraceLangChainHandler. Install it with: pip install langchain-core"
            )
        self._tracker = RunTracker()
        self._last_llm_run_id: str | None = None

    # ── LLM callbacks ────────────────────────────────────────────────────

    def on_chat_model_start(
        self,
        serialized: dict[str, Any],
        messages: list[list[Any]],
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> None:
        model_name = (metadata or {}).get("ls_model_name", "")
        provider = (metadata or {}).get("ls_provider", "")
        input_data = _serialize_messages(messages[0]) if messages else None

        self._tracker.on_run_start(
            name=f"{provider}.chat" if provider else "chat_model",
            run_id=str(run_id),
            parent_run_id=str(parent_run_id) if parent_run_id else None,
            kind="llm",
            input=input_data,
            model=model_name,
            provider=provider,
            attributes={
                "etrace.kind": "llm",
                "gen_ai.system": provider,
                "gen_ai.request.model": model_name,
            },
        )
        self._last_llm_run_id = str(run_id)

    def on_llm_start(
        self,
        serialized: dict[str, Any],
        prompts: list[str],
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> None:
        model_name = (metadata or {}).get("ls_model_name", "")
        provider = (metadata or {}).get("ls_provider", "")

        self._tracker.on_run_start(
            name=f"{provider}.chat" if provider else "llm",
            run_id=str(run_id),
            parent_run_id=str(parent_run_id) if parent_run_id else None,
            kind="llm",
            input=prompts[0] if prompts else None,
            model=model_name,
            provider=provider,
            attributes={
                "etrace.kind": "llm",
                "gen_ai.system": provider,
                "gen_ai.request.model": model_name,
            },
        )
        self._last_llm_run_id = str(run_id)

    def on_llm_end(
        self,
        response: Any,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> None:
        pending = self._tracker._spans.get(str(run_id))
        if not pending:
            return

        usage: dict[str, int] = {}
        output: Any = None
        extra_attrs: dict[str, Any] = {}

        gens = response.generations[0][0] if response.generations else None
        if gens:
            msg = getattr(gens, "message", None)
            if msg:
                rm = getattr(msg, "response_metadata", {})
                tu = rm.get("token_usage", {})
                prompt = tu.get("prompt_tokens", 0) or 0
                completion = tu.get("completion_tokens", 0) or 0
                if prompt or completion:
                    usage = {
                        "prompt_tokens": prompt,
                        "completion_tokens": completion,
                        "total_tokens": tu.get("total_tokens", prompt + completion),
                    }

                content = getattr(msg, "content", None)
                if content:
                    output = content
                    extra_attrs["gen_ai.output"] = str(content)[:100_000]
                else:
                    tool_calls = getattr(msg, "tool_calls", None)
                    if tool_calls:
                        tc_data = [
                            {
                                "name": tc.get("name", ""),
                                "arguments": tc.get("args", {}),
                            }
                            for tc in tool_calls
                        ]
                        output = json.dumps(tc_data)
                        extra_attrs["gen_ai.output"] = output

                resp_model = rm.get("model_name", "")
                if resp_model:
                    extra_attrs["gen_ai.response.model"] = resp_model

        self._tracker.on_run_end(
            str(run_id),
            output=output,
            usage=usage or None,
            model=extra_attrs.get("gen_ai.response.model"),
            attributes=extra_attrs or None,
        )

    def on_llm_error(
        self,
        error: BaseException,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> None:
        self._tracker.on_run_error(str(run_id), error)

    # ── Tool callbacks ────────────────────────────────────────────────────

    def on_tool_start(
        self,
        serialized: dict[str, Any],
        input_str: str,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        inputs: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> None:
        tool_name = serialized.get("name", "tool")

        # In LangGraph, the tools node is a sibling of the model node,
        # so parent_run_id points to "tools" chain, not the LLM.
        # We track the last LLM run_id so tools become children of
        # the LLM that triggered them.
        effective_parent = self._last_llm_run_id or parent_run_id

        self._tracker.on_run_start(
            name=tool_name,
            run_id=str(run_id),
            parent_run_id=str(effective_parent) if effective_parent else None,
            kind="tool",
            input=inputs or input_str,
            attributes={"etrace.kind": "tool"},
        )

    def on_tool_end(
        self,
        output: Any,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> None:
        self._tracker.on_run_end(str(run_id), output=output)

    def on_tool_error(
        self,
        error: BaseException,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> None:
        self._tracker.on_run_error(str(run_id), error)

    # ── Chain callbacks ───────────────────────────────────────────────────

    def on_chain_start(
        self,
        serialized: dict[str, Any],
        inputs: dict[str, Any],
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> None:
        # Record the run_id so children can resolve their parent, but
        # do NOT create an etrace span.  LangGraph chain nodes are
        # internal plumbing (middleware, routing) that add noise.
        self._tracker._spans[str(run_id)] = _Run(
            span=None,
            start=0.0,
            parent_run_id=str(parent_run_id) if parent_run_id else None,
        )

    def on_chain_end(
        self,
        outputs: dict[str, Any],
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> None:
        # Noop — chain runs are not traced, only used for parent lookup.
        pass

    def on_chain_error(
        self,
        error: BaseException,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> None:
        pass

    # ── Lifecycle ─────────────────────────────────────────────────────────

    def flush(self) -> None:
        """Close any remaining open runs."""
        self._tracker.flush()

File: sdks/typescript/src/langchain.ts
/**
 * Callback adapter for framework-driven agent tracing.
 *
 * The class intentionally avoids importing the optional framework package at
 * module load time. Callback managers use method names on callback objects, so
 * this duck-typed handler can be passed directly in config.callbacks.
 */
import { MAX_ATTR_LEN } from "./index.js";
import { RunTracker } from "./tracing.js";
import type { UsageMap } from "./tracing.js";

export function langchainHandler(): EtraceLangChainHandler {
  return new EtraceLangChainHandler();
}

export class EtraceLangChainHandler {
  name = "EtraceLangChainHandler";

  private readonly _tracker = new RunTracker();
  private _lastLlmRunId: string | null = null;

  handleChatModelStart(
    llm: unknown,
    messages: unknown[][],
    runId: string,
    parentRunId?: string,
    _extraParams?: unknown,
    _tags?: string[],
    metadata?: Record<string, unknown>,
    _runName?: string,
  ): void {
    const provider = stringValue(metadata?.ls_provider) ?? stringValue(metadata?.provider) ?? "";
    const model = stringValue(metadata?.ls_model_name) ?? stringValue(metadata?.model) ?? "";

    this._tracker.onRunStart(provider ? `${provider}.chat` : "chat_model", {
      runId,
      parentRunId,
      kind: "llm",
      input: messages?.[0] ? serializeMessages(messages[0]) : undefined,
      model,
      provider,
      attributes: {
        "etrace.kind": "llm",
        "gen_ai.system": provider,
        "gen_ai.request.model": model,
      },
    });
    this._lastLlmRunId = runId;
  }

  handleLLMStart(
    _llm: unknown,
    prompts: string[],
    runId: string,
    parentRunId?: string,
    _extraParams?: unknown,
    _tags?: string[],
    metadata?: Record<string, unknown>,
    _runName?: string,
  ): void {
    const provider = stringValue(metadata?.ls_provider) ?? stringValue(metadata?.provider) ?? "";
    const model = stringValue(metadata?.ls_model_name) ?? stringValue(metadata?.model) ?? "";

    this._tracker.onRunStart(provider ? `${provider}.chat` : "llm", {
      runId,
      parentRunId,
      kind: "llm",
      input: prompts?.[0],
      model,
      provider,
      attributes: {
        "etrace.kind": "llm",
        "gen_ai.system": provider,
        "gen_ai.request.model": model,
      },
    });
    this._lastLlmRunId = runId;
  }

  handleLLMEnd(response: unknown, runId: string): void {
    const { output, usage, model, attributes } = parseLlmResponse(response);
    this._tracker.onRunEnd(runId, {
      output,
      usage,
      model,
      attributes,
    });
  }

  handleLLMError(error: unknown, runId: string): void {
    this._tracker.onRunError(runId, error);
  }

  handleToolStart(
    tool: unknown,
    input: string,
    runId: string,
    parentRunId?: string,
    _tags?: string[],
    _metadata?: Record<string, unknown>,
    _runName?: string,
    inputs?: Record<string, unknown>,
  ): void {
    const effectiveParent = this._lastLlmRunId ?? parentRunId;
    this._tracker.onRunStart(toolName(tool), {
      runId,
      parentRunId: effectiveParent,
      kind: "tool",
      input: inputs ?? input,
      attributes: { "etrace.kind": "tool" },
    });
  }

  handleToolEnd(output: unknown, runId: string): void {
    this._tracker.onRunEnd(runId, { output });
  }

  handleToolError(error: unknown, runId: string): void {
    this._tracker.onRunError(runId, error);
  }

  handleChainStart(
    _chain: unknown,
    _inputs: Record<string, unknown>,
    runId: string,
    parentRunId?: string,
  ): void {
    this._tracker.skipRun(runId, parentRunId);
  }

  handleChainEnd(_outputs: Record<string, unknown>, _runId: string): void {
    // Chain runs are used only for parent lookup; they are intentionally quiet.
  }

  handleChainError(_error: unknown, _runId: string): void {
    // Chain runs have no etrace span to close.
  }

  flush(): void {
    this._tracker.flush();
  }
}

function serializeMessages(messages: unknown[]): Array<Record<string, unknown>> {
  return messages.map((message) => {
    const msg = asRecord(message);
    const entry: Record<string, unknown> = {
      role: stringValue(msg.type) ?? getMessageType(msg) ?? stringValue(msg.role) ?? "unknown",
      content: msg.content ?? "",
    };
    const toolCalls = asArray(msg.tool_calls ?? msg.toolCalls);
    if (toolCalls.length) {
      entry.tool_calls = toolCalls.map((toolCall) => serializeToolCall(toolCall));
    }
    return entry;
  });
}

function parseLlmResponse(response: unknown): {
  output?: unknown;
  usage?: UsageMap | null;
  model?: string;
  attributes?: Record<string, unknown>;
} {
  const res = asRecord(response);
  const generation = firstGeneration(res.generations);
  const message = asRecord(asRecord(generation).message);
  const responseMetadata = asRecord(message.response_metadata ?? message.responseMetadata);
  const llmOutput = asRecord(res.llmOutput ?? res.llm_output);
  const tokenUsage = asRecord(
    responseMetadata.token_usage ??
      responseMetadata.tokenUsage ??
      llmOutput.tokenUsage ??
      llmOutput.token_usage,
  );

  const usage = parseUsage(tokenUsage);
  const attributes: Record<string, unknown> = {};
  const model =
    stringValue(responseMetadata.model_name) ??
    stringValue(responseMetadata.modelName) ??
    stringValue(llmOutput.model_name) ??
    stringValue(llmOutput.modelName);
  if (model) attributes["gen_ai.response.model"] = model;

  let output: unknown;
  const content = message.content ?? asRecord(generation).text;
  if (content) {
    output = content;
    attributes["gen_ai.output"] = stringOrJson(content).slice(0, MAX_ATTR_LEN);
  } else {
    const toolCalls = asArray(message.tool_calls ?? message.toolCalls);
    if (toolCalls.length) {
      const serialized = toolCalls.map((toolCall) => serializeToolCall(toolCall));
      output = serialized;
      attributes["gen_ai.output"] = JSON.stringify(serialized).slice(0, MAX_ATTR_LEN);
    }
  }

  return {
    output,
    usage,
    model,
    attributes: Object.keys(attributes).length ? attributes : undefined,
  };
}

function firstGeneration(generations: unknown): unknown {
  const outer = asArray(generations);
  const first = outer[0];
  const inner = asArray(first);
  return inner.length ? inner[0] : first;
}

function parseUsage(tokenUsage: Record<string, unknown>): UsageMap | null {
  const prompt =
    numberValue(tokenUsage.prompt_tokens) ??
    numberValue(tokenUsage.promptTokens) ??
    numberValue(tokenUsage.input_tokens) ??
    0;
  const completion =
    numberValue(tokenUsage.completion_tokens) ??
    numberValue(tokenUsage.completionTokens) ??
    numberValue(tokenUsage.output_tokens) ??
    0;
  const total =
    numberValue(tokenUsage.total_tokens) ??
    numberValue(tokenUsage.totalTokens) ??
    prompt + completion;
  const cached =
    numberValue(tokenUsage.cached_tokens) ??
    numberValue(tokenUsage.cachedTokens) ??
    numberValue(asRecord(tokenUsage.prompt_tokens_details).cached_tokens) ??
    0;
  const reasoning =
    numberValue(tokenUsage.reasoning_tokens) ??
    numberValue(tokenUsage.reasoningTokens) ??
    numberValue(asRecord(tokenUsage.completion_tokens_details).reasoning_tokens) ??
    0;

  if (!prompt && !completion && !total && !cached && !reasoning) return null;
  return {
    prompt_tokens: prompt,
    completion_tokens: completion,
    total_tokens: total,
    cached_tokens: cached,
    reasoning_tokens: reasoning,
  };
}

function serializeToolCall(toolCall: unknown): Record<string, unknown> {
  const tc = asRecord(toolCall);
  return {
    name: tc.name ?? asRecord(tc.function).name ?? "",
    arguments: tc.args ?? tc.arguments ?? asRecord(tc.function).arguments ?? {},
  };
}

function toolName(tool: unknown): string {
  const obj = asRecord(tool);
  return stringValue(obj.name) ?? stringValue(asRecord(obj.id).name) ?? "tool";
}

function getMessageType(message: Record<string, unknown>): string | undefined {
  const getType = message._getType;
  if (typeof getType !== "function") return undefined;
  try {
    return stringValue(getType.call(message));
  } catch {
    return undefined;
  }
}

function stringOrJson(value: unknown): string {
  if (typeof value === "string") return value;
  try {
    return JSON.stringify(value);
  } catch {
    return String(value);
  }
}

function asRecord(value: unknown): Record<string, unknown> {
  return value !== null && typeof value === "object" ? (value as Record<string, unknown>) : {};
}

function asArray(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

function stringValue(value: unknown): string | undefined {
  return typeof value === "string" && value.length ? value : undefined;
}

function numberValue(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}
